### PR TITLE
backend: make changes to accounts atomic

### PR DESCRIPTION
**Pull request by benma**
We had an accountsLock, but it not held for the whole duration of
initAccounts(), which means that in concurrent calls, the accounts
could be uninitialized twice, then all accounts would be added twice,
or other such race conditions.

This commit only acquires lock in exported functions to make it work
with all combinations of account-changing function calls.

### CODE REVIEW

File: backend/backend.go [hunk 1]
```diff
@@ -420,7 +432,7 @@ func (backend *Backend) createAndAddBTCAccount(
 			case signing.ScriptTypeP2WPKH:
 				suffixedName += ": bech32"
 			}
-			err := backend.CreateAndAddAccount(
+			err := backend.createAndAddAccount(
```
Comment (x1ddos) [inline backend/backend.go:435]: Please add it to `createAndAddBTCAccount` too that it requires the accountsLock.

File: backend/backend.go [hunk 2]
```diff
@@ -1155,7 +1181,10 @@ func (backend *Backend) Close() error {
 	errors := []string{}
 
 	backend.ratesUpdater.Stop()
+
+	unlock := backend.accountsLock.Lock()
```
Comment (x1ddos) [inline backend/backend.go:1185]: I would just do it at the top with defer'ed unlock. After all, it says: "After this, no other method should be called."

Might be irrelevant but I noticed this manipulates `backend.coins`. I think it should also locked for the whole duration and possibly `delete(backend.coins, coin)`?

File: backend/backend.go [hunk 3]
```diff
@@ -228,7 +228,6 @@ func (backend *Backend) configureHistoryExchangeRates() {
 
 // addAccount adds the given account to the backend.
 func (backend *Backend) addAccount(account accounts.Interface) {
```
Comment (x1ddos) [inline backend/backend.go:230]: please add the note about accountsLock

**Review by x1ddos — COMMENTED** [review]
`func (backend *Backend) Accounts() []accounts.Interface` also needs a lock. Otherwise, you get a concurrent usage race condition for `accounts` field, unless I misunderstand what `accountLock` is for?

But locking in `Backend.Accounts` wouldn't work because `initAccounts` requires a lock now and it calls `configureHistoryExchangeRates`, which in turn calls `Backend.Accounts`. Deadlock.

File: backend/backend.go (hunk 1 above)
Comment (benma) [reply backend/backend.go:440]: done

**Comment by benma** [comment]
> `func (backend *Backend) Accounts() []accounts.Interface` also needs a lock. Otherwise, you get a concurrent usage race condition for `accounts` field, unless I misunderstand what `accountLock` is for?
> 
> But locking in `Backend.Accounts` wouldn't work because `initAccounts` requires a lock now and it calls `configureHistoryExchangeRates`, which in turn calls `Backend.Accounts`. Deadlock.

Great catch. Fixed.

File: backend/backend.go (hunk 3 above)
Comment (benma) [reply backend/backend.go:233]: done

File: backend/backend.go (hunk 2 above)
Comment (benma) [reply backend/backend.go]: done (not the coins part, maybe in another PR)

File: backend/backend.go [hunk 4]
```diff
@@ -208,9 +208,11 @@ func NewBackend(arguments *arguments.Arguments, environment Environment) (*Backe
 
 // configureHistoryExchangeRates changes backend.ratesUpdater settings.
 // It requires both backend.config to be up-to-date and all accounts initialized.
+//
+// The accountsLock must be held when calling this function.
 func (backend *Backend) configureHistoryExchangeRates() {
 	var coins []string
-	for _, acct := range backend.Accounts() {
+	for _, acct := range backend.accounts {
```
Comment (x1ddos) [inline backend/backend.go:215]: You got off easy this time :)

File: backend/backend.go [hunk 5]
```diff
@@ -852,7 +877,10 @@ func (backend *Backend) Start() <-chan interface{} {
 	if backend.arguments.DevMode() {
 		backend.baseManager.Start()
 	}
+
+	unlock := backend.accountsLock.Lock()
 	backend.initPersistedAccounts()
+	unlock()
```
Comment (x1ddos) [inline backend/backend.go:883]: but now that configureHistoryExchangeRates requires the accontsLock, this unlock should be at the very end past configureHistory call. I would `defer` it.

**Review by x1ddos — APPROVED** [review]
LGTM with one more place to look at.

File: backend/backend.go (hunk 5 above)
Comment (benma) [reply backend/backend.go]: damnit! fixed :)

**Pull request merged**